src/api/attribute/service.ts

```diff
@@ -0,0 +1,160 @@
+
+import request from 'request';
+import TagCache from 'redis-tag-cache'
+import get from 'lodash/get';
+import cache from '../../lib/cache-instance'
+
+export interface AttributeListParam {
+  [key: string]: number[]
+}
+
+/**
+ * Build ES uri fro attributes
+ */
+function getUri (config, indexName: string): string {
+  return `${config.elasticsearch.protocol}://${config.elasticsearch.host}:${config.elasticsearch.port}/${indexName}/attribute/_search`
+}
+
+/**
+ * Returns attributes from cache
+ */
+async function getAttributeFromCache (attributeCode: string, config) {
+  if (config.server.useOutputCache && cache) {
+    try {
+      const res = await (cache as TagCache).get(
+        'api:attribute-list' + attributeCode
+      )
+      return res
+    } catch (err) {
+      console.error(err)
+      return null
+    }
+  }
+}
+
+/**
+ * Save attributes in cache
+ */
+async function setAttributeInCache (attributeList, config) {
+  if (config.server.useOutputCache && cache) {
+    try {
+      await Promise.all(
+        attributeList.map(attribute => (cache as TagCache).set(
+          'api:attribute-list' + attribute.attribute_code,
+          attribute
+        ))
+      )
+    } catch (err) {
+      console.error(err)
+    }
+  }
+}
+
+/**
+ * Returns attribute with only needed options
+ * @param attribute - attribute object
+ * @param optionsIds - list of only needed options ids
+ */
+function clearAttributeOpitons (attribute, optionsIds: number[]) {
+  const stringOptionsIds = optionsIds.map(String)
+  return {
+    ...attribute,
+    options: (attribute.options || []).filter(option => stringOptionsIds.includes(String(option.value)))
+  }
+}
+
+function list (attributesParam: AttributeListParam, config, indexName: string): Promise<any[]> {
+  return new Promise(async (resolve, reject) => {
+    // we start with all attributeCodes that are requested
+    let attributeCodes = Object.keys(attributesParam)
+
+    // here we check if some of attribute are in cache
+    const rawCachedAttributeList = await Promise.all(
+      attributeCodes.map(attributeCode => getAttributeFromCache(attributeCode, config))
+    )
+
+    const cachedAttributeList = rawCachedAttributeList
+      .filter(Boolean) // remove empty results from cache.get
+      .map((cachedAttribute, index) => {
+        if (cachedAttribute) {
+          const attributeOptionsIds = attributesParam[cachedAttribute.attribute_code]
+
+          // side effect - we want to reduce starting 'attributeCodes' because some of them are in cache
+          attributeCodes.splice(index, 1)
+
+          // clear unused options
+          return clearAttributeOpitons(cachedAttribute, attributeOptionsIds)
+        }
+      })
+
+    // if all requested attributes are in cache then we can return here
+    if (!attributeCodes.length) {
+      return resolve(cachedAttributeList)
+    }
+
+    // fetch attributes for rest attributeCodes
+    request({
```

Comment:
we could have used the `elasticsearch-js` client, but it should work that way as well - I guess

Reply:
done :)